### PR TITLE
Pun it/429 android bugs

### PR DESCRIPTION
**Pull request by Pun-it**

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
### Summary
This PR improves resilience around connectivity, normalizes phone matching, and fixes DTO parsing/serialization inconsistencies. The biggest remaining risks are around unintended SSE reconnection fan-out and a regression in pagination handling that can silently truncate valid content lists.

### Issues

#### HIGH — Pagination loop can terminate early when API legitimately returns items without IDs
- File: `Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt`
- Function: anonymous coroutine inside content loading block (`viewModelScope.launch`)
- Problem:
  The new deduplication logic filters out all content items with `id == null` and immediately breaks the pagination loop when `newItems.isEmpty()`. If the backend returns a page containing valid content entries without IDs (or duplicate IDs during pagination overlap), pagination stops even when `nextCursor` still points to additional pages.
- Impact:
  Incomplete classroom content loading with silent data loss in UI selection flows.
- Fix:
  Continue pagination based on cursor semantics instead of `newItems.isEmpty()`. Deduplicate only when IDs are present:
  - retain items with null IDs
  - only skip duplicates for non-null IDs
  - remove the early `break`

#### MEDIUM — Reconnection method can create duplicate SSE subscriptions
- File: `Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt`
- Function: `reconnectSSEIfNeeded`
- Problem:
  `startSSE(confId)` is called unconditionally whenever connectivity flips to ONLINE. There is no guard ensuring an existing SSE stream is closed or already inactive before reconnecting.
- Impact:
  Multiple active SSE listeners can lead to duplicated participant updates, inconsistent UI state, and unnecessary network/socket usage.
- Fix:
  Track SSE connection state and avoid reconnecting if an active stream already exists, or explicitly close/dispose the previous stream before calling `startSSE`.

#### MEDIUM — `strip91` normalization incorrectly alters non-Indian numbers beginning with `91`
- File: `Teacher-App/app/src/main/java/com/example/seeds/adapters/StudentCallStatusAdapter.kt`
- Function: `strip91`
- Problem:
  The normalization removes any leading `"91"` when length exceeds 10, regardless of whether the value is actually an Indian country-code-prefixed number.
- Impact:
  Incorrect student matching for international or malformed numbers that naturally begin with `91`.
- Fix:
  Normalize only E.164-style Indian numbers:
  - strip optional `"+"`
  - verify total normalized length matches expected Indian mobile format before removing country code
  - centralize this logic in a shared phone utility instead of duplicating lambdas/functions across files

#### LOW — Repeated phone normalization logic creates divergence risk
- File: `Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt`
- Function: `handleSSEUpdate`
- Problem:
  A second inline normalization lambda duplicates the new `strip91` logic added elsewhere.
- Impact:
  Future normalization fixes can diverge between adapter and SSE reconciliation behavior.
- Fix:
  Extract a single shared phone normalization utility and reuse it in both adapter and ViewModel logic.

### Score
- HIGH: 1 × -3 = -3
- MEDIUM: 2 × -2 = -4
- LOW: 1 × -1 = -1

Final Score: 92/100

**Review by a4i-architect — COMMENTED** [review]
### Summary
This PR improves phone normalization, classroom DTO compatibility, connectivity handling, and SSE recovery logic. The overall direction is good, but there are a few correctness regressions around pagination termination, SSE reconnection duplication, and DTO serialization compatibility that can break production behavior.

### Issues

#### HIGH — `CallViewModel.kt` / `loadContentForClassroom` can terminate pagination early and silently drop content
- File: `Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt`
- Function: `loadContentForClassroom` (inside added pagination loop)

The new deduplication logic breaks pagination whenever a page contains only previously-seen IDs. In cursor-based APIs this can legitimately happen due to overlapping windows or unstable ordering, causing the loop to stop before later pages are fetched.

Current behavior:
- `newItems.isEmpty()` immediately breaks the loop even if `nextCursor` points to additional unseen pages.

Concrete fix:
- Remove the `if (newItems.isEmpty()) break` condition.
- Continue paging until `nextCursor == null`.
- Keep deduplication only for insertion into `allContentList`.

Severity rationale:
- This can cause incomplete classroom content loading for users in production.

---

#### HIGH — `CallViewModel.kt` / `reconnectSSEIfNeeded` can create duplicate SSE streams
- File: `Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt`
- Function: `reconnectSSEIfNeeded`

`startSSE(confId)` is invoked unconditionally whenever connectivity returns ONLINE, without checking whether an SSE connection is already active or reconnecting. Repeated ONLINE events can create parallel SSE subscriptions, duplicated participant updates, and leaked network resources.

Concrete fix:
- Track SSE connection state/job explicitly.
- Return early if an active SSE stream already exists.
- Cancel any existing SSE collector before starting a new one.

Severity rationale:
- Duplicate realtime streams can corrupt UI state and degrade app/network performance.

---

#### MEDIUM — `ClassroomDto.kt` / removal of `@ToJson` breaks backward-compatible request serialization
- File: `Teacher-App/app/src/main/java/com/example/seeds/network/ClassroomDto.kt`
- Function: `ClassroomDto.Companion`

The custom `@ToJson` adapter was removed while the DTO field naming changed from `_id` to `id`. If any existing API endpoints or cached payloads still expect `_id`, Moshi default serialization will now emit `id`, potentially breaking updates or partial compatibility with older backend contracts.

Concrete fix:
- Restore a dedicated `@ToJson` adapter or annotate fields with explicit `@Json(name = "_id")` / dual-model mapping to guarantee stable wire format compatibility.

Severity rationale:
- This introduces request/response contract drift that can break classroom save/update flows depending on backend deployment state.

---

#### MEDIUM — `StudentCallStatusAdapter.kt` / `strip91` normalization is inconsistent with formatted phone numbers
- File: `Teacher-App/app/src/main/java/com/example/seeds/adapters/StudentCallStatusAdapter.kt`
- Function: `strip91`

The normalization only strips a leading `91` but does not handle common formats such as `+91XXXXXXXXXX`, whitespace, or separators. This can still fail student matching and DiffUtil identity comparisons after this change.

Concrete fix:
- Normalize by removing non-digits first, then strip leading country code consistently.
- Centralize normalization into a shared utility used by both adapter and ViewModel SSE merge logic.

Severity rationale:
- Causes inconsistent participant matching and unstable RecyclerView updates for valid phone inputs.

---

### Score
- HIGH: 2 × -3 = -6
- MEDIUM: 2 × -2 = -4

Final Score: 90/100

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallFragment.kt [hunk 1]
```diff
@@ -300,6 +300,7 @@ class CallFragment : BaseFragment() {
             }
             ConnectivityStatus.ONLINE -> {
                 container.visibility = View.GONE
+                viewModel.reconnectSSEIfNeeded()
```
Comment (farmanahmed888) [inline Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallFragment.kt]: we should not do reconnectSSEIfNeeded, it should connect on first try, when we called it

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt [hunk 2]
```diff
@@ -460,6 +461,13 @@ class CallViewModel @Inject constructor(
         }
     }
 
+    fun reconnectSSEIfNeeded() {
+        val confId = _callToken.value?.confId ?: return
+        conferenceEverRunning.set(false)
+        Log.i(TAG, "SSE: Reconnecting after connectivity recovery")
+        startSSE(confId)
+    }
```
Comment (farmanahmed888) [inline Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt]: not needed, start SSE initially should be enough

**Review by farmanahmed888 — CHANGES_REQUESTED** [review]
see following comments, and make changes if needed

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt (hunk 2 above)
Comment (N1kunj1998) [inline Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt]: `conferenceEverRunning` guards the "navigate teacher back when call ends" logic. Resetting to `false` before reconnect means the very first SSE event with `is_running=false` (sent by server after a blip) immediately navigates the teacher out of an ongoing call.

**Failure scenario:** Network drops and recovers mid-call → teacher reconnects → server sends `is_running=false` heartbeat → teacher is bounced to home screen while students are still in the call.

**Fix:**
```kotlin
fun reconnectSSEIfNeeded() {
    val confId = _callToken.value?.confId ?: return
    // Do NOT reset conferenceEverRunning — the conference was running before the blip
    Log.i(TAG, "SSE: Reconnecting after connectivity recovery")
    startSSE(confId)
}
```

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt [hunk 3]
```diff
@@ -517,7 +528,16 @@ class CallViewModel @Inject constructor(
...
+            // Preserve pre-answer students (RINGING) not yet in SSE participants.
+            // SSE only includes participants who have joined; a student still ringing
+            // externally won't appear until they answer, so we keep them in current state.
+            val normalize = { p: String -> if (p.startsWith("91") && p.length > 10) p.substring(2) else p }
+            val sseKeys = students.mapNotNull { it.phoneNumber?.let(normalize) }.toSet()
+            val preAnswerStudents = (_callState.value ?: emptyList()).filter { existing ->
+                val key = existing.phoneNumber?.let(normalize) ?: return@filter false
+                key !in sseKeys && participantTrackers[existing.phoneNumber]?.hasBeenInCall != true
+            }
+            _callState.postValue((students + preAnswerStudents).sortedByDescending { it.raiseHand })
```
Comment (N1kunj1998) [inline Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt]: `sseKeys` is normalized; `participantTrackers` lookup is raw. A student with `participantTrackers["919876543210"]` won't be found via `existing.phoneNumber = "9876543210"` → `hasBeenInCall` always null → student appears in both `students` and `preAnswerStudents` → duplicate row every SSE tick.

**Failure scenario:** Student answers call → still shows as both RINGING and CONNECTED in teacher UI simultaneously.

**Fix:**
```kotlin
val preAnswerStudents = (_callState.value ?: emptyList()).filter { existing ->
    val key = existing.phoneNumber?.let(normalize) ?: return@filter false
    val trackerKey = participantTrackers.keys.firstOrNull { normalize(it) == key }
    key !in sseKeys && participantTrackers[trackerKey]?.hasBeenInCall != true
}
```

File: Teacher-App/app/src/main/java/com/example/seeds/adapters/StudentCallStatusAdapter.kt [hunk 4]
```diff
@@ -10,6 +10,11 @@ import com.example.seeds.model.StudentCallStatus
 import com.example.seeds.model.Student
 import com.example.seeds.ui.call.CallViewModel
 
+private fun strip91(p: String?): String {
+    val s = p ?: return ""
+    return if (s.startsWith("91") && s.length > 10) s.substring(2) else s
```
Comment (N1kunj1998) [inline Teacher-App/app/src/main/java/com/example/seeds/adapters/StudentCallStatusAdapter.kt]: Any 11-digit number whose first two digits are `9` and `1` (e.g. local number `91XXXXXXXXX`) gets truncated to 9 digits, matching nobody. Same logic duplicated inline in `CallViewModel` as `normalize` lambda.

**Failure scenario:** Student whose real local number is `91XXXXXXXX` (starts with 91) → stripped to 9 digits → no match in `allStudents` → synthetic fallback Student shown with `_id=null`.

**Fix:** Check for exactly 12-digit E.164 format instead:
```kotlin
private fun strip91(p: String?): String {
    val s = p ?: return ""
    return if (s.length == 12 && s.startsWith("91")) s.substring(2) else s
}
```
Apply the same fix to the `normalize` lambda in `CallViewModel.kt`.

**Review by a4i-architect — COMMENTED** [review]
### Summary
This PR improves phone normalization, pagination safety, DTO parsing flexibility, and connectivity handling. Most changes are sound, but there are two correctness regressions introduced around DTO serialization compatibility and content pagination behavior that can break production flows.

### Issues

#### HIGH — `Teacher-App/app/src/main/java/com/example/seeds/network/ClassroomDto.kt` — `ClassroomSaveDto`
Changing the serialized field from `_id` to `id` in `ClassroomSaveDto` is a breaking API contract change unless the backend was updated simultaneously. Existing APIs commonly expect `_id` for Mongo-backed resources, and this DTO no longer includes any Moshi annotations to preserve backward compatibility.

Concrete fix:
- Restore `_id` as the serialized field name, or add `@Json(name = "_id")` to the `id` property.
- Add/update a serialization unit test validating the outbound JSON payload for classroom save requests.

#### HIGH — `Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt` — content pagination loop inside `viewModelScope.launch`
The new deduplication logic silently drops all content items whose `id` is `null`. Previously these items were still included. If the backend legitimately returns content without IDs, users will no longer see that content.

Concrete fix:
- Preserve null-ID content items instead of filtering them out.
- Deduplicate only non-null IDs, e.g. keep entries where `id == null` and dedupe only when an ID exists.
- Add a unit test covering mixed null/non-null content IDs across paginated responses.

#### MEDIUM — `Teacher-App/app/src/main/java/com/example/seeds/network/ClassroomDto.kt` — `ClassroomDto.Companion.fromJson`
The new `"id", "_id"` handling calls `reader.nextString()` directly. If either field is explicitly `null` in API responses, Moshi will throw `JsonDataException` and fail deserialization.

Concrete fix:
- Guard nullable values using `reader.peek()` before `nextString()`, or use `nextNull<Unit>()` handling when the token is `NULL`.
- Add a deserialization test for payloads containing `"id": null` and `"_id": null`.

### Score
95/100

Deductions:
- HIGH: 2 × -3 = -6
- MEDIUM: 1 × -2 = -2

Total: 100 - 8 = 92

Correction after sanity check:
- The pagination null-ID issue depends on backend behavior and may not always break functionality. Downgraded from HIGH to MEDIUM.

Final Score: 94/100

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt (hunk 2 above)
Comment (Pun-it) [reply Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt]: Removed `reconnectSSEIfNeeded()` entirely. `ConferenceSSEClient` builds its EventSource with `reconnectTime(3s)`, so the stream auto-reconnects on connectivity loss. Initial `startSSE()` on callToken is enough.
**Confirmed physically with a phone.**

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallFragment.kt (hunk 1 above)
Comment (Pun-it) [reply Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallFragment.kt]: Dropped the reconnect call from the ONLINE branch; SSE connects once on first try (startSSE when callToken arrives) and the EventSource library handles reconnection itself.

File: Teacher-App/app/src/main/java/com/example/seeds/adapters/StudentCallStatusAdapter.kt (hunk 4 above)
Comment (Pun-it) [reply Teacher-App/app/src/main/java/com/example/seeds/adapters/StudentCallStatusAdapter.kt]: Fixed `strip91` to strip the country code only for exact E.164 India form (12 digits = 91 + 10).
Applied the same fix to the `normalize` lambda in CallViewModel.handleSSEUpdate.

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt (hunk 2 above)
Comment (Pun-it) [reply Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt]: Resolved by removing reconnectSSEIfNeeded() entirely (per farman's comment too). The EventSource is built with reconnectTime(3s) so it auto-reconnects — the manual reconnect + conferenceEverRunning reset was redundant and caused exactly the mid-call bounce you described. The reset code no longer exists; conferenceEverRunning is still set true in handleSSEUpdate so the call-ended navigation is intact.

File: Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt (hunk 3 above)
Comment (Pun-it) [reply Teacher-App/app/src/main/java/com/example/seeds/ui/call/CallViewModel.kt]: Couldn't reproduce a visible duplicate in manual testing, but the raw-vs-normalized key mismatch is real: trackers are keyed in server (E.164) form while existing.phoneNumber can be local, so participantTrackers[existing.phoneNumber] can miss and read hasBeenInCall as null. Fixed defensively by matching on the normalized key so it can't cause a duplicate row later.

**Review by N1kunj1998 — APPROVED** [review]
LGTM

**Review by farmanahmed888 — APPROVED** [review]
LGTM
